pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,325 @@
+	return queue
+}
+
+// NewCommand creates a command key and adds in initial data for the orchestration queue.
+func NewCommand(replacements []nodeclaim.Key, candidates []*state.StateNode, reason string, timeAdded time.Time) *Command {
+	return &Command{
+		ReplacementKeys: lo.Map(replacements, func(key nodeclaim.Key, _ int) *NodeClaimKey {
+			return &NodeClaimKey{Key: key}
+		}),
+		Candidates: candidates,
+		Reason:     reason,
+		TimeAdded:  timeAdded,
+	}
+}
+
+// Start will kick off the queue. This runs asynchronously unil Karpenter shuts down.
+func (q *Queue) Start(ctx context.Context) {
+	for {
+		// Get command from queue. This waits until queue is non-empty.
+		item, shutdown := q.RateLimitingInterface.Get()
+		if shutdown {
+			break
+		}
+		hash := item.(uint64)
+		cmd := q.idToCommand[hash]
+		requeue, err := q.Handle(ctx, cmd)
+		if !requeue {
+			// If the command timed out, log the last error.
+			if err != nil {
+				// If the command failed, bail on the action.
+				// 1. Emit metrics for launch failures
+				// 2. Ensure cluster state no longer thinks these nodes are deleting
+				// 3. Remove it from the Queue's internal data structure
+				deprovisioningReplacementNodeLaunchFailedCounter.WithLabelValues(cmd.Reason).Inc()
+				q.cluster.UnmarkForDeletion(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string { return s.ProviderID() })...)
+				err = multierr.Append(err, q.setNodesUnschedulable(ctx, false, lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string { return s.Node.Name })...))
+				logging.FromContext(ctx).Debugf("handling deprovisioning command, %w, for nodes %s", err,
+					strings.Join(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+						return s.Name()
+					}), ","))
+			}
+			q.Remove(hash, cmd)
+			continue
+		}
+		q.RateLimitingInterface.Done(hash)
+		// Requeue command if not complete
+		q.RateLimitingInterface.AddRateLimited(hash)
+	}
+	logging.FromContext(ctx).Errorf("deprovisioning queue is broken and has shutdown")
+}
+
+// Add adds commands to the Queue
+// Each command added to the queue should already be validated and ready for execution.
+func (q *Queue) Add(cmd *Command) error {
+	providerIDs := lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+		return s.ProviderID()
+	})
+	// First check if we can add the command.
+	if err := q.CanAdd(providerIDs...); err != nil {
+		return fmt.Errorf("adding command, %w", err)
+	}
+	hash, err := Hash(cmd)
+	if err != nil {
+		return err
+	}
+	for _, candidate := range cmd.Candidates {
+		q.candidateProviderIDToCommandID[candidate.ProviderID()] = hash
+		q.idToCommand[hash] = cmd
+	}
+	// Idempotently mark for deletion
+	q.cluster.MarkForDeletion(providerIDs...)
+	q.RateLimitingInterface.Add(hash)
+	return nil
+}
+
+// Remove fully clears the queue of all references of a hash/command
+func (q *Queue) Remove(hash uint64, cmd *Command) {
+	q.RateLimitingInterface.Forget(hash)
```

Comment:
nit: stylistic thing but I generally think it's nice to keep calls to related or the same structs/interfaces together. If I see them broken out like this, I tend to think that there's some sort of reason that they were broken out this way (race condition, etc.).